zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,289 @@
+            .filter_map(|input| match input {
+                TransparentInput::PrevOut { outpoint, .. } => Some(outpoint),
+                TransparentInput::Coinbase { .. } => None,
+            })
+            .try_for_each(|outpoint| outpoint.zcash_serialize(&mut buf))?;
+
+        hash.update(&buf);
+
+        writer.write_all(hash.finalize().as_ref())
+    }
+
+    fn hash_sequence<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        if self.hash_type & SIGHASH_ANYONECANPAY != 0
+            || (self.hash_type & SIGHASH_MASK) == SIGHASH_SINGLE
+            || (self.hash_type & SIGHASH_MASK) == SIGHASH_NONE
+        {
+            return writer.write_all(&[0; 32]);
+        }
+
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(ZCASH_SEQUENCE_HASH_PERSONALIZATION)
+            .to_state();
+
+        let mut buf = vec![];
+
+        self.trans
+            .inputs()
+            .map(|input| match input {
+                TransparentInput::PrevOut { sequence, .. } => sequence,
+                TransparentInput::Coinbase { sequence, .. } => sequence,
+            })
+            .try_for_each(|sequence| (&mut buf).write_u32::<LittleEndian>(*sequence))?;
+
+        hash.update(&buf);
+
+        writer.write_all(hash.finalize().as_ref())
+    }
+
+    /// Writes the u256 hash of the transactions outputs to the provided Writer
+    fn hash_outputs<W: io::Write>(&self, writer: W) -> Result<(), io::Error> {
+        todo!()
+        // if (self.hash_type & SIGHASH_MASK) == SIGHASH_SINGLE
+        //     || (self.hash_type & SIGHASH_MASK) == SIGHASH_NONE
+        // {
+        //     self.outputs_hash(writer)
+        // } else if (self.hash_type & SIGHASH_MASK) == SIGHASH_SINGLE
+        // //     && transparent_input.is_some()
+        // //     && transparent_input.as_ref().unwrap().0 < tx.vout.len()
+        // {
+        //     self.single_output_hash(writer)
+        // } else {
+        //     writer.write_all(&[0; 32])
+        // }
+    }
+
+    fn outputs_hash<W: io::Write>(&self, writer: W) -> Result<(), io::Error> {
+        todo!()
+    }
+
+    fn single_output_hash<W: io::Write>(&self, writer: W) -> Result<(), io::Error> {
+        todo!()
+    }
+
+    fn hash_joinsplits<W: io::Write>(&self, writer: W) -> Result<(), io::Error> {
+        todo!()
+
+        // if !tx.joinsplits.is_empty() {
+        //     writer
+        //         .write_all(&[0; 32])
+        //         .expect("write to hasher will never fail");
+        // }
+
+        // todo!()
+    }
+
+    fn hash_lock_time<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        self.trans.lock_time().zcash_serialize(&mut writer)
+    }
+
+    fn hash_expiry_height<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_u32::<LittleEndian>(
+            self.trans
+                .expiry_height()
+                .expect("fOverwintered is always set")
```

Comment:
Question: What stops this code being run on a v1 or v2 transaction?

```suggestion
                .expect("Transaction is V3 or later")
```